specs/phase1/data-availability-sampling.md

```diff
@@ -0,0 +1,218 @@
+```
+
+#### `reverse_bit_order_list`
+
+```python
+def reverse_bit_order_list(elements: Sequence[int]) -> Sequence[int]:
+    order = len(elements)
+    assert is_power_of_two(order)
+    return [elements[reverse_bit_order(i, order)] for i in range(order)]
+```
+
+### Data extension
+
+Implementations:
+- [Python](https://github.com/protolambda/partial_fft/blob/master/das_fft.py)
+- [Go](https://github.com/protolambda/go-kate/blob/master/das_extension.go)
+
+```python
+def das_fft_extension(data: Sequence[Point]) -> Sequence[Point]:
+    """
+    Given some even-index values of an IFFT input, compute the odd-index inputs,
+    such that the second output half of the IFFT is all zeroes.
+    """
+    poly = inverse_fft(data)
+    return fft(poly + [0]*len(poly))[1::2]
+```
+
+### Data recovery
+
+See [Reed-Solomon erasure code recovery in n*log^2(n) time with FFTs](https://ethresear.ch/t/reed-solomon-erasure-code-recovery-in-n-log-2-n-time-with-ffts/3039) for theory.
+Implementations:
+- [Original Python](https://github.com/ethereum/research/blob/master/mimc_stark/recovery.py)
+- [New optimized approach in python](https://github.com/ethereum/research/tree/master/polynomial_reconstruction)
+- [Old approach in Go](https://github.com/protolambda/go-kate/blob/master/recovery.go)
+
+```python
+def recover_data(data: Sequence[Optional[Sequence[Point]]]) -> Sequence[Point]:
+    """Given an a subset of half or more of subgroup-aligned ranges of values, recover the None values."""
+    ...
+```
+
+## DAS functions
+
+```python
+def extend_data(data: Sequence[Point]) -> Sequence[Point]:
+    """
+    The input data gets reverse-bit-ordered, such that the first half of the final output matches the original data.
+    We calculated the odd-index values with the DAS FFT extension, reverse-bit-order to put them in the second half.
+    """
+    rev_bit_odds = reverse_bit_order_list(das_fft_extension(reverse_bit_order_list(data)))
+    return data + rev_bit_odds
+```
+
+```python
+def unextend_data(extended_data: Sequence[Point]) -> Sequence[Point]:
+    return extended_data[:len(extended_data)//2]
+```
+
+```python
+def check_multi_kate_proof(commitment: BLSCommitment, proof: BLSKateProof, x: Point, ys: Sequence[Point]) -> bool:
+    """
+    Run a KZG multi-proof check to verify that for the subgroup starting at x,
+    the proof indeed complements the ys to match the commitment.
+    """
+    ...  # Omitted for now, refer to Kate implementation resources.
+```
+
+```python
+def construct_proofs(extended_data_as_poly: Sequence[Point]) -> Sequence[BLSKateProof]:
+    """
+    Constructs proofs for samples of extended data (in polynomial form, 2nd half being zeroes).
+    Use the FK20 multi-proof approach to construct proofs for a chunk length of POINTS_PER_SAMPLE.
+    """
+    ... # Omitted for now, refer to Kate implementation resources.
+```
+
+```python
+def commit_to_data(data_as_poly: Sequence[Point]) -> Sequence[BLSCommitment]:
```

Comment:
```suggestion
def commit_to_data(data_as_poly: Sequence[Point]) -> BLSCommitment:
```

Reply:
Thanks, copied the signature from the proofs function and forgot about the return type.